rpc/src/v1/impls/eth.rs

```diff
@@ -718,28 +750,32 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM, T: StateInfo + 'static> Eth for EthClient<
 		};
 
 		let transaction_id = PendingTransactionId::Location(block_id, index.value());
-		Box::new(future::done(self.transaction(transaction_id)))
+		let result = self.transaction(transaction_id).and_then(
+			errors::check_block_number_existence(&*self.client, num, self.options.allow_missing_blocks));
+		Box::new(future::done(result))
 	}
 
 	fn transaction_receipt(&self, hash: RpcH256) -> BoxFuture<Option<Receipt>> {
+		let best_block = self.client.chain_info().best_block_number;
 		let hash: H256 = hash.into();
 
-		if self.options.allow_pending_receipt_query {
-			let best_block = self.client.chain_info().best_block_number;
-			if let Some(receipt) = self.miner.pending_receipt(best_block, &hash) {
-				return Box::new(future::ok(Some(receipt.into())));
+		match (self.miner.pending_receipt(best_block, &hash), self.options.allow_pending_receipt_query) {
```

Comment:
This will unnecessarily call `self.miner.pending_receipt` even when `self.options.allow_pending_receipt_query` is false. I would suggest to change it back to the old two `if` form.